hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationManager.java

```diff
@@ -216,10 +217,19 @@ void finalizeMigration(MigrationInfo migrationInfo) {
 
                 MigrationEndpoint endpoint = source ? MigrationEndpoint.SOURCE : MigrationEndpoint.DESTINATION;
                 FinalizeMigrationOperation op = new FinalizeMigrationOperation(migrationInfo, endpoint, success);
-
                 op.setPartitionId(partitionId).setNodeEngine(nodeEngine).setValidateTarget(false)
                         .setService(partitionService);
-                nodeEngine.getOperationService().execute(op);
+                InternalOperationService operationService = nodeEngine.getOperationService();
+                if (operationService.isRunAllowed(op)) {
+                    // When migration finalization is triggered by subsequent migrations
+                    // on partition thread, finalization may run directly on calling thread.
+                    // Otherwise, if previously completed migration and newly submitted migration
```

Comment:
If I understood correctly, this was about the partition thread queue containing multiple migrations where we wanted to both process the migration and finalize atomically, right? Instead, `execute` would place the finalize operation at the end of the partition queue.

Reply:
Yes. This only happens when a source of a former migration becomes destination of a latter migration for the same partition. 

During migration commit, destination of the migration learns the commit via direct invocation but other members, including the source, learn the commit when they receive completed migrations. Completed migrations are sent via migration operations to ensure that source and destination apply them before processing the new migration. If completed migrations include a migration for the same partition, it must be finalized before processing the new one. That's why we run finalization on the same thread if it belongs to the same partition thread. It may belong to a different partition belonging to same thread, in that case, it doesn't matter whether it's run now or later.

For example; this member is `MemberA`. It was owning replica `0` of a partition `P`. 
- `Migration-1` migrates replica `0` of partition `P` from `MemberA` to `MemberB`. 
- `Migration-2` migrates replica `3` of partition `P` from `MemberC` to `MemberA`. 

So, `MemberA` is source of `Migration-1` and destination of `Migration-2`. For `MemberA`, 
- `Migration-1` means `remove data for partition=P replica=0`
- `Migration-2` means `write data for partition=P replica=3`

When `MemberA` receives migration operation for `Migration-2`, it must finalize `Migration-1` before processing `Migration-2`. Otherwise, finalization of `Migration-1` will delete data copied by `Migration-2`.

Reply:
Thanks for the explanation!